mindmeld/components/entity_resolver.py

```diff
@@ -42,60 +46,284 @@
+        cls.validate_resolver_name(name)
+        return ENTITY_RESOLVER_MODEL_MAPPINGS.get(name)(
+            app_path, resource_loader, entity_type, er_config, **kwargs
+        )
+
+    def fit(self):
+        raise NotImplementedError
+
+    def load(self):
+        raise NotImplementedError
+
+    def predict(self, entity):
+        raise NotImplementedError
+
+
+class EntityResolverBase(ABC):
+    """
+    Base class for Entity Resolvers
+    """
+
+    def __init__(self, app_path, resource_loader, entity_type, er_config, **kwargs):
+        """Initializes an entity resolver"""
+        self.app_path = app_path
         self.type = entity_type
+        self.er_config = er_config
+        self.kwargs = kwargs
+
+        self._app_namespace = get_app_namespace(self.app_path)
         self._is_system_entity = Entity.is_system_entity(self.type)
-        self._exact_match_mapping = None
-        self._er_config = get_classifier_config("entity_resolution", app_path=app_path)
-        self._es_host = es_host
-        self._es_config = {"client": es_client, "pid": os.getpid()}
-        self.ready = False
+        self.name = self.er_config.get("model_type")
+        self.entity_map = resource_loader.get_entity_map(self.type)
+        self.normalizer = resource_loader.query_factory.normalize
 
-        if self._is_system_entity:
-            canonical_entities = []
-        else:
-            canonical_entities = self._resource_loader.get_entity_map(self.type).get(
-                "entities", []
-            )
-        self._no_canonical_entity_map = len(canonical_entities) == 0
+        if self._use_double_metaphone:
+            self._enable_double_metaphone()
 
-    @property
-    def _es_index_name(self):
-        return EntityResolver.ES_SYNONYM_INDEX_PREFIX + "_" + self.type
+        self.dirty = False  # bool, True if exists any unsaved generated data that can be saved
+        self.ready = False  # bool, True if the model is fit by calling .fit()
 
     @property
-    def _use_text_rel(self):
-        return self._er_config["model_type"] == "text_relevance"
+    def _use_double_metaphone(self):
+        return "double_metaphone" in self.er_config.get("phonetic_match_types", [])
+
+    def _enable_double_metaphone(self):
+        """
+        By default, resolvers are assumed to not support double metaphone usage
+        If supported, override this method definition in the derived class
+        (eg. see ElasticsearchEntityResolver)
+        """
+        logger.warning(
+            "%r not configured to use double_metaphone",
+            self.name
+        )
+        raise NotImplementedError
+
+    def cache_path(self, tail_name=""):
+        name = f"{self.name}_{tail_name}" if tail_name else self.name
+        return path.get_entity_resolver_cache_file_path(
+            self.app_path, self.type, name
+        )
+
+    @abstractmethod
+    def _fit(self, clean):
+        raise NotImplementedError
+
+    def fit(self, clean=False):
```

Comment:
We need to update the comment about the `clean` argument. Also i wonder we probably should re-visit the usage of `clean`. With the new Bert based implementation maybe we should do things that are similar in other models where we have a configurable option for re-training the model when the data is changed (i.e. entity mapping for resolver) instead of requiring the developer to explicit specify `clean` to re-train

Reply:
Got it. Just to confirm, you suggestion is to (automatically) identify if the data got changed and decide whether to do a re-training or not? Because we are storing synonyms (whitelists plus canonical names) and their embeddings in a pickle file, we can obtain BERT embeddings for any newly added synonyms and update the stored pickle file. Does that sound right?

Reply:
Yeah exactly - we could detect the data and config change to build new model or load dumped model accordingly so that developers won't have to keep track of it and set `clean` flag. (to not load stale model after data/config change)

I think the hash idea is great. In fact I believe we're doing something very similar for classifier models. In the `fit()` of base `Classifier` class  we generate hash based on queries, model config and other resources [(here)](https://github.com/cisco/mindmeld/blob/4054b84239ddbf00fceabab303b3f2687c053354/mindmeld/components/classifier.py#L547) and build new model when data or config has changed if the incremental build flag is on.

Reply:
@mhuang2173 I just committed a new version of code with this functionality. I also updated this PR's description to reflect that.